packages/editor/src/ui-tools/tool-bar/tool-bar.tsx

```diff
@@ -86,83 +79,31 @@ export class ToolBar extends GLSPAbstractUIExtension implements IActionHandler,
     this.toDisposeOnHide.dispose();
   }
 
-  protected createHeader(): void {
-    const headerCompartment = createElement('div', ['tool-bar-header']);
-    headerCompartment.appendChild(this.createLeftButtons());
-    headerCompartment.appendChild(this.createMiddleButtons());
-    headerCompartment.appendChild(this.createRightButtons());
-    this.containerElement.innerHTML = '';
-    this.containerElement.appendChild(headerCompartment);
-    this.changeActiveButton();
-  }
-
-  private createLeftButtons(): HTMLElement {
-    const leftButtons = createElement('div', ['left-buttons']);
-
-    this.defaultToolsButton = this.createToolButton(DefaultSelectButton);
-    leftButtons.appendChild(this.defaultToolsButton);
-    leftButtons.appendChild(this.createToolButton(MarqueeToolButton));
-
-    if (!this.editorContext.isReadonly) {
-      const editPart = createElement('div', ['edit-buttons']);
-      editPart.appendChild(this.createToolButton(UndoToolButton));
-      editPart.appendChild(this.createToolButton(RedoToolButton));
-      leftButtons.appendChild(editPart);
-    }
-    return leftButtons;
-  }
-
-  private createMiddleButtons(): HTMLElement {
-    const middleButtons = createElement('div', ['middle-buttons']);
-    this.appendProvidedButtons(middleButtons, ToolBarButtonLocation.Middle);
-    return middleButtons;
-  }
-
-  private createRightButtons(): HTMLElement {
-    const rightButtons = createElement('div', ['right-buttons']);
-    this.appendProvidedButtons(rightButtons, ToolBarButtonLocation.Right);
-    return rightButtons;
-  }
-
-  private appendProvidedButtons(parent: HTMLElement, location: ToolBarButtonLocation): void {
-    this.toolBarButtonProvider
+  protected render(): React.ReactNode {
+    return (
+      <ToolBarComponent
```

Comment:
as on the viewport bar, I think it makes not so much sense to pass this buttons as props, I prefer to use the JSX syntax and pass such things as children.

Reply:
Here, it is a bit different because we do have four different kind of buttons essentially (left, edit, middle, right). I'll see if there is a nicer way of handling this but otherwise if we are just doing children, then the toolbar UI extension needs to specify the whole structure as it has the necessary GLSP context.